add support for debian packages

pazos opened the pull request:
Fixes #3108 and **maybe** #4398 

Requires https://github.com/koreader/koreader-base/pull/785

Frenzie commented on platform/debian/do_debian_package.sh:
```diff
@@ -0,0 +1,62 @@
+PACKAGE="KOReader"
+AUTHOR="KOReader community"
+DESC_1="An ebook reader application supporting PDF, DjVu, EPUB, FB2 and many more formats"
+DESC_2="https://github.com/koreader/koreader"
```
Probably better to use:
```suggestion
DESC_2="https://koreader.rocks"
```

pazos replied:
updated homepage :+1:

Frenzie commented on platform/debian/do_debian_package.sh:
```diff
@@ -0,0 +1,62 @@
+    } >"$CONTROL"
+    dpkg-deb -b "${PKG_PATH}" "${ROOTFS_BASE}/${FULL_NAME}"
+else
+    echo "${COMMAND} not found, skipping target ${FULL_NAME}"
```
Shouldn't you exit 1 here?

Frenzie commented on platform/debian/koreader.sh:
```diff
@@ -0,0 +1,36 @@
+# we're always starting from our working directory
+cd "${KOREADER_DIR}" || exit
+
+# export load library path for some old firmware
```
This is only for some old Pocketbooks really; I probably more or less accidentally added it to the AppImage since I partially based it on the Ubuntu Touch port (while I suppose you based this on the AppImage).

See https://github.com/koreader/koreader/commit/327624e4fa1dd6bab765142995095905cd54c41a

pazos replied:
ok, on my machine it works if I remove `export LD_LIBRARY_PATH` but I see that everywhere. https://github.com/koreader/koreader/search?q=export+LD_LIBRARY_PATH&unscoped_q=export+LD_LIBRARY_PATH

Frenzie replied:
Good point, it might otherwise fail to resolve the correct libraries (i.e., ours). Looks like it was added as a precaution in https://github.com/koreader/koreader/commit/bb164c1deb87a7e0ac411e52eef293b915c6e978

pazos commented:
updated a bunch of things:

1. add common mimetypes on .desktop
2. package as root and move datastorage to $HOME/.config/koreader
3. beautifully do_debian_package.sh, now reports estimated size and the right homepage
4. if arch is not specified asume native and try to get debian arch names from `uname -m`. Fallback is any for essoteric archs.
5. add runtime depends on libsdl2

pazos commented:
I had to move installation from /usr/local/ to /usr to make the desktop file able to open files using free-desktop standard. This has a few benefits (dpkg won't display warnings when updating/erasing the package, graphical installers will have the nice KOReader icon)

I also moved the subpath from share/koreader to lib/koreader, since it is the default for all-in-one applications (like firefox).

Last, but not least, I removed duplicated fonts (they're in fonts/ and in resources/fonts) to reduce the installation size to ~80MB and the package itself to ~27MB. I guess we should do the same on the appimage front.

Frenzie commented:
> Last, but not least, I removed duplicated fonts (they're in fonts/ and in resources/fonts) to reduce the installation size to ~80MB and the package itself to ~27MB. I guess we should do the same on the appimage front.

Good catch! :-)

Frenzie commented:
Btw, there are still some Lintian errors that can be easily fixed. But I figure it's more interesting to have it merged already right now.

Most or all of this for example, most notably for the short term the package name and copyright file:
```
E: KOReader: missing-dependency-on-libc needed by usr/lib/koreader/common/libluacompat52.so and 40 others
E: KOReader: debian-changelog-file-missing
E: KOReader: no-copyright-file
W: KOReader: description-synopsis-starts-with-article
W: KOReader: description-too-long
W: KOReader: extended-description-line-too-long
W: KOReader: extended-description-line-too-long
E: KOReader: bad-package-name
E: KOReader: package-not-lowercase
E: KOReader: maintainer-address-missing KOReader team
W: KOReader: duplicate-font-file usr/lib/koreader/fonts/freefont/FreeSans.ttf also in fonts-freefont-ttf
W: KOReader: duplicate-font-file usr/lib/koreader/fonts/freefont/FreeSerif.ttf also in fonts-freefont-ttf
W: KOReader: duplicate-font-file usr/lib/koreader/fonts/noto/NotoSans-Bold.ttf also in fonts-noto-hinted
W: KOReader: duplicate-font-file usr/lib/koreader/fonts/noto/NotoSans-BoldItalic.ttf also in fonts-noto-hinted
W: KOReader: duplicate-font-file usr/lib/koreader/fonts/noto/NotoSans-Italic.ttf also in fonts-noto-hinted
W: KOReader: duplicate-font-file usr/lib/koreader/fonts/noto/NotoSans-Regular.ttf also in fonts-noto-hinted
W: KOReader: duplicate-font-file usr/lib/koreader/fonts/noto/NotoSerif-Bold.ttf also in fonts-noto-hinted
W: KOReader: duplicate-font-file usr/lib/koreader/fonts/noto/NotoSerif-BoldItalic.ttf also in fonts-noto-hinted
W: KOReader: duplicate-font-file usr/lib/koreader/fonts/noto/NotoSerif-Italic.ttf also in fonts-noto-hinted
W: KOReader: duplicate-font-file usr/lib/koreader/fonts/noto/NotoSerif-Regular.ttf also in fonts-noto-hinted
W: KOReader: package-contains-vcs-control-file usr/lib/koreader/jit/.gitignore
W: KOReader: binary-without-manpage usr/bin/koreader
```

As an aside, I was slightly confused by these:
```
E: KOReader: unstripped-binary-or-object usr/lib/koreader/common/libluacompat52.so
```
The release version isn't supposed to be a debug build, so did Lintian effectively find a bug? It'll have to be investigated somewhat. :-)

Anyway, thanks!

pazos commented:
@Frenzie: I didn't know lintian. Ooops..

> E: KOReader: missing-dependency-on-libc needed ...

Yes, we probably need to specify that as a build-depends extracting libc6-dev version from the system where the package is built. If we want to release the package as nightlies then we want to build on the oldest debian-based system available.

> W: KOReader: duplicate-font-file ...

Not sure how to fix that. Font path seems hardcoded.

> E: KOReader: maintainer-address-missing KOReader team

Yep, maintainer field expects: Name of Maintainer <email@domain.com>

> E: KOReader: debian-changelog-file-missing
E: KOReader: no-copyright-file
W: KOReader: description-synopsis-starts-with-article
W: KOReader: description-too-long
W: KOReader: extended-description-line-too-long
W: KOReader: extended-description-line-too-long
E: KOReader: bad-package-name
E: KOReader: package-not-lowercase

easy to fix :+1: 

Will check myself w/ lintian output

Frenzie commented:
manpage is also easy to fix. We can basically just stick this in there:

https://github.com/koreader/koreader/blob/6de5927dc27bd1af5c8361ca48064806eb481e5e/reader.lua#L45-L59

> Not sure how to fix that. Font path seems hardcoded.

You could symlink them in an install script or something, but those are harder and simultaneously less important issues imho.

> If we want to release the package as nightlies then we want to build on the oldest debian-based system available.

I'd say the same Ubuntu 16.04 already used for Docker should do just fine. ;-)

zaoqi-unsafe commented:
When can I download prebuilt *.deb?
